### PR TITLE
chore: remove eol python 3.7

### PR DESCRIPTION
**Pull request by devkral**
Remove python 3.7 and add python 3.11 to remaining test matrices 

This PR also updates the testrunner docker image to python 3.8 and fully qualifies the source

## Types of Changes

- [ ] Core
- [x] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage). (maximal damage: broken tests)

### CODE REVIEW

**Comment by botberry** [comment]
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/), so make sure to pick the appropriate type. If in doubt feel free to ask :)


Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to Alexander for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```

**Comment by devkral** [comment]
this PR also updates the dependencies (required because of python update)

**Comment by kristjanvalur** [comment]
If 3.7 is dropped, then the special handling of `asyncio.CancelledError` can be removed, since it becomes a direct child of  `BaseException` with 3.8

**Comment by devkral** [comment]
as far as I can see, asyncio.CancelledError are not special handled (only for cleanup they are suppressed but for a good reason)

**Comment by kristjanvalur** [comment]
> as far as I can see, asyncio.CancelledError are not special handled (only for cleanup they are suppressed but for a good reason)

Code like this:
```python
        try:
            ...
        except asyncio.CancelledError:
            raise
        except Exception as error:
            await self.handle_task_exception(error)  # pragma: no cover
        
```
is written for 3.7 compatibility, where `asyncio.CancelledError` is a subclass of `Exeption` yet should not be handled but re-raised.
In 3.8 it is a direct `BaseException` and the above clause can be omitted (because there is no `except BaseException` anywhere in sight)

**Comment by devkral** [comment]
as said:

asyncio.CancelledError is propagated

the other errors are handled in:

handle_task_exception

It is quite a change of the logic if you handle now asyncio.CancelledError in handle_task_exception

**Comment by devkral** [comment]
is something like 
``` patch
--- a/strawberry/subscriptions/protocols/graphql_transport_ws/handlers.py
+++ b/strawberry/subscriptions/protocols/graphql_transport_ws/handlers.py
@@ -108,8 +108,6 @@ class BaseGraphQLTransportWSHandler(ABC):
             self.connection_timed_out = True
             reason = "Connection initialisation timeout"
             await self.close(code=4408, reason=reason)
-        except asyncio.CancelledError:
-            raise
         except Exception as error:
             await self.handle_task_exception(error)  # pragma: no cover
         finally:
@@ -118,7 +116,10 @@ class BaseGraphQLTransportWSHandler(ABC):
             self.completed_tasks.append(task)
 
     async def handle_task_exception(self, error: Exception) -> None:
-        self.task_logger.exception("Exception in worker task", exc_info=error)
+        if isinstance(error, asyncio.CancelledError):
+            self.task_logger.exception("Worker task cancelled", exc_info=error)
+        else
+            self.task_logger.exception("Exception in worker task", exc_info=error)
 
     async def handle_message(self, message: dict) -> None:
         handler: Callable

```
right?
Handling CancelledErrors  in handle_task_exception is quite a change

**Comment by kristjanvalur** [comment]
Just the first part. In 3.8 CancelledError won't be caught by "except Exception" because it isn't an "Exception"

File: strawberry/subscriptions/protocols/graphql_transport_ws/handlers.py [hunk 1]
```diff
@@ -116,14 +116,16 @@ async def handle_connection_init_timeout(self) -> None:
...
-        except Exception as error:
+        except BaseException as error:
+            # asyncio.CancelledError don't inherit from Exception
             await self.handle_task_exception(error)  # pragma: no cover
         finally:
             # do not clear self.connection_init_timeout_task
             # so that unittests can inspect it.
             self.completed_tasks.append(task)
 
-    async def handle_task_exception(self, error: Exception) -> None:
```
Comment (kristjanvalur) [inline strawberry/subscriptions/protocols/graphql_transport_ws/handlers.py:120]: This change should be removed.

File: strawberry/subscriptions/protocols/graphql_transport_ws/handlers.py [hunk 2]
```diff
@@ -116,14 +116,16 @@ async def handle_connection_init_timeout(self) -> None:
...
+    async def handle_task_exception(self, error: BaseException) -> None:
+        # asyncio.CancelledError don't inherit from Exception
         if isinstance(error, asyncio.CancelledError):
```
Comment (kristjanvalur) [inline strawberry/subscriptions/protocols/graphql_transport_ws/handlers.py:122]: This code can be reverted.

**Comment by kristjanvalur** [comment]
I think I may have confused you with my talk about "CancelledError".  Basically, we never want to catch it.
Previously, for 3.7, to ensure that we didn't, we had to add in `except asyncio.CancelledError: raise` clauses, in order to avoid catching it.   for 3.8 and above, this is no longer necessary, as long as you are not catching `BaseException` (which you should almost never do).

So, the only thing required, is to simply __drop__ the `except asyncio.CancelledError: raise` clauses

**Comment by devkral** [comment]
> I think I may have confused you with my talk about "CancelledError". Basically, we never want to catch it. Previously, for 3.7, to ensure that we didn't, we had to add in `except asyncio.CancelledError: raise` clauses, in order to avoid catching it. for 3.8 and above, this is no longer necessary, as long as you are not catching `BaseException` (which you should almost never do).
> 
> So, the only thing required, is to simply **drop** the `except asyncio.CancelledError: raise` clauses

n some new python versions asyncio.CancelledError inherits from Exception, so we need it

**Comment by devkral** [comment]
cleaned up and rebased.

Also add the updated dependencies in the release.md

**Comment by kristjanvalur** [comment]
> in some new python versions asyncio.CancelledError inherits from Exception, so we need it

This is incorrect.  as the documentation clearly states:  "Changed in version 3.8: [CancelledError](https://docs.python.org/3/library/asyncio-exceptions.html?highlight=cancellederror#asyncio.CancelledError) is now a subclass of [BaseException](https://docs.python.org/3/library/exceptions.html#BaseException)."

**Comment by devkral** [comment]
technically this stanca would be also right, if they inherit from Exception, so it is implementation specific

**Comment by kristjanvalur** [comment]
> technically this stanca would be also right, if they inherit from Exception, so it is implementation specific

No it isn't.  Yes, you could _deliberately_ mis-interpret it, but it indicates a _change_ from the 3.7 behaviour.  It is not implementation specific.  I don't quite understand why you are being so stubborn about this, it is a well known fact of Python asyncio programming.   If you like, I can have it fixed in the documentation.

**Comment by devkral** [comment]
Thing is: vscode says something different. Maybe the stubs are wrong.

If you say it works, then I give it a try.

I just don't want to break anything as I have not so much time with two children

**Comment by devkral** [comment]
this too?
````python
        except asyncio.CancelledError:
            # CancelledErrors are expected during task cleanup.
            pass
```

**Comment by kristjanvalur** [comment]
It is fine, I can clean this all up later, I just wanted to mention this in the PR since this was about deprecating 3.7 and I have been wanting to get this out of the way for a long time.

Fwiw, I made a change to the docs, and here you have it from the horse's mouth:
https://github.com/python/cpython/pull/106453#pullrequestreview-1514751783

**Comment by maximyurevich** [comment]
Also uvicorn can be updated with editing this PR

**Comment by patrick91** [comment]
@devkral can you rebase this? 😊

**Comment by devkral** [comment]
done

**Comment by devkral** [comment]
I hope it is ok to add an editorconfig as my defaults always wreck the yaml formatting.

**Comment by devkral** [comment]
there were some new refs to python 3.7 which had to be removed (tests/pyright).

Hope we can end the never ending PR soon ;)

**Review by patrick91 — APPROVED** [review]
Let's do this 😊

**Pull request merged**